docs/prover/checking/sanity.md

```diff
@@ -2,9 +2,241 @@ Rule Sanity Checks
+
+    ```solidity
+     contract sanityCheck{
+     address root;
+     uint a;
+     constructor(address _root){
+         require(_root != address(0));
+         root = _root;
+     }
+     function changeRoot (address root_) public{
+         require(root_ != address(0));
+         root = root_;
+     }
+     }
+     ```
+   
+   
+    ```cvl
+    
+      // Not a tautology
+      invariant rootNonZero()
+         root() != 0
+
+      rule rootNonZero(){
+         assert root() != 0;
+      }
+
+      // Tautology
+      invariant aGE0()
+         a() >= 0
+
+      rule aGE0Rule{
+         assert a() >= 0;
+      }
+
+    ```
+    
+    **Checking vacuity for rules**
+    
+    For rules, checking for tautology requires checking each assertion to see if 
+    it’s meaningful. In order to do this, we employ few different checks depending
+    on the syntax of the assertion expression.
+    
+      *Case 1: Implication*
+    
+      Given a rule with an `assert p => q` we perform two checks:
+    
+      1. Implication hypothesis: `assert(!p)`
+       
+      If the hypothesis part is always false then the assertion is a tautology.
+      
+      ```cvl
+              rule testSanity{
+              uint a;
+              uint b;
+              assert a<0 => b<10;
+              }
+      ```
+         
+        
+      Error message
+        
+      ```cvl
+            assert-vacuity check FAILED: sanity.spec:11:5
+            assert-tautology check FAILED: sanity.spec:11:5'a < 0 => b < 10' is a vacuous 
+            implication. It could be rewritten to !a < 0 because a < 0 is always false
+            
+      ```
+      
+        
+    2. Implication conclusion: `assert(q)`
+       
+       If the conclusion part is always true regardless of the hypothesis then the
+       assertion is a tautology
+       
+        ```cvl
+          rule testSanity{
+          uint a;
+          uint b;
+          assert a>10 => b>=0;
+          }
+        ```
+        
+      Error Message
+        
+      ```cvl
+              assert-tautology check FAILED: sanity.spec:21:5conclusion `b >= 0` is always true 
+              regardless of the hypothesis
+      ```
+        
+   *Case 2: Double Implication*
+     
+     Given a rule with an assert p <=> q we perform two checks:
+     
+     1. Double implication, both false: `assert(!p && !q)`
+         If this passes then the assertion is a tautology since both conditions are always false.
```

Comment:
```suggestion
         If this passes then the assertion is a tautology since both conditions are always false leaving no way for the biconditional statement to evaluate to false.
```